docs/Blog-Post.md

```diff
@@ -0,0 +1,525 @@
+
+# Named Axes for Tensor Dimensions
+Pos = hax.Axis("position", 1024)  # sequence
+KPos = Pos.alias("key_position")  # key sequence for attention
+Head = hax.Axis("head", 8)  # number of attention heads
+Key = hax.Axis("key", 64)  # key/query/value size
+Embed = hax.Axis("embed", 512)  # embedding size
+
+def attention(Key, KPos, query, key, value, mask):
+    # how similar is each query to each key
+    scores = hax.dot(Key, query, key) / jnp.sqrt(Key.size)
+
+    # mask out invalid positions
+    if mask is not None:
+      scores -= 1E9 * (1.0 - mask)
+
+    # convert to probabilities
+    scores = hax.nn.softmax(scores, axis=KPos)
+
+    # weighted sum of values
+    return hax.dot(KPos, scores, value)
+```
+
+In this example, we've defined an axis for each dimension of our tensors. In Haliax, the named `Axis` is the basic
+building block of named tensors, pairing a name with a size. We can then use these axes to define our tensors, and use
+those axes to perform operations like `softmax` and tensor multiplication (`dot`).
+
+## Compositionality
+
+Despite making no reference to batching or heads, this same implementation is also batch-capable and supports multi-headed
+(or multi-query) attention and even attending to or from non-sequential keys (e.g. attending to image patches):
+
+```python
+Batch = hax.Axis("batch", 8)  # batch size
+
+query = hax.random.normal(PRNGKey(0), (Batch, Head, Pos, Key))
+key = hax.random.normal(PRNGKey(1), (Batch, Head, KPos, Key))
+value = hax.random.normal(PRNGKey(2), (Batch, Head, KPos, Key))
+
+# traditional batched multi-headed attention
+assert attention(Key, KPos, query, key, value, mask=None).axes == (Batch, Head, Pos, Key)
+
+# multi-query attention. Each key/value pair produces only one head
+key = hax.random.normal(PRNGKey(1), (Batch, KPos, Key))
+value = hax.random.normal(PRNGKey(2), (Batch, KPos, Key))
+assert attention(Key, KPos, query, key, value, mask=None).axes == (Batch, Head, Pos, Key)
+
+# image patch cross-attention from a sequence
+Height = hax.Axis("height", 32)
+Width = hax.Axis("width", 32)
+
+key = hax.random.normal(PRNGKey(1), (Batch, Head, Height, Width, Key))
+value = hax.random.normal(PRNGKey(2), (Batch, Head, Height, Width, Key))
+
+# KPos in attention actually be a tuple of axes.
+assert attention(Key, (Height, Width), query, key, value, mask=None).axes == (Batch, Head, Pos, Key)
+```
+
+This compositionality is possible because we've abstracted over the unreferenced dimensions of our tensors.
+In the first example, both the `Batch` and `Head` axes are unreferenced, so they are automatically "batched" over.
+Similarly, in the second example, we omit the `Head` axis from the `key` and `value` tensors, but attention still works.
+In the third example, we can use tuples of axes in many places where we would normally use a single axis.
+
+
+## Avoiding Bugs
+
+Earlier, I claimed that named tensors can help avoid common bugs. Here's an example of a bug that is easy to make
```

Comment:
Not sure we should use the first-person singular in general...

Reply:
fair enough. i'll redo